proximal/halide/src/least_square_direct.cpp

```diff
@@ -121,9 +127,27 @@ class least_square_direct_gen : public Generator<least_square_direct_gen> {
             ;
 
         if(!ignore_offset) {
-            f_offset //
-            .compute_root()
-            .parallel(c);
+            // Note(Antony): We cache the Fourier domain of image "offset",
+            // likely the input raw pixel image. Since the raw image is a
+            // constant through out the ADMM iterations, caching avoids having
+            // to recompute the Fourier domain again.
+
+            // The cache-to-recompute tradeoff is data size dependent. For four
+            // mega-pixel (4 MP) worth of data, we normally obtain around 8%
+            // improvement with caching on a 12-core computer. Image smaller
+            // than 1 MP may benefit from recomputation as the naive approach.
+            f_offset_cached  //
+                .compute_root()
+                .bound(k, 0, 2)
+                .unroll(k)
+                .vectorize(x, vfloat)
+                .parallel(y)
+                .parallel(c)
+                .memoize();
+
+            // Ensure the Halide generated FFT subroutine is inside the
+            // condition "is_cache_miss".
+            f_offset_tmp.compute_at(f_offset_cached, c);
```

Comment:
Tagging @soufianekhiat . Checking if the benefit of caching outweighs the cost of heap memory overhead.

Reply:
Pretty cool for this PR.
I need to profile on my side.